umbral/signing.py

```diff
@@ -1,151 +1,143 @@
+        return cls(r, s), data
 
-        return cls(CurveBN.from_int(r, curve), CurveBN.from_int(s, curve))
+    def __bytes__(self):
+        return bytes(self.r) + bytes(self.s)
 
-    def _der_encoded_bytes(self) -> bytes:
-        return utils.encode_dss_signature(int(self.r), int(self.s))
+    def __eq__(self, other):
+        return self.r == other.r and self.s == other.s
 
-    def __bytes__(self) -> bytes:
-        return self.r.to_bytes() + self.s.to_bytes()
 
-    def __len__(self):
-        return len(bytes(self))
+def signature_recovery_flag(message: bytes,
+                            signature: Signature,
+                            public_key: PublicKey,
+                            ) -> bool:
+    """
+    Returns the recovery value of a signature applied to `message` and verifiable by `public_key`.
+    """
 
-    def __add__(self, other):
-        return bytes(self) + other
+    prehashed_message = digest_for_signing(message).finalize()
 
-    def __radd__(self, other: bytes) -> bytes:
-        return other + bytes(self)
+    r = signature.r
+    s = signature.s
 
-    def __eq__(self, other) -> bool:
-        simple_bytes_match = hmac.compare_digest(bytes(self), bytes(other))
-        der_encoded_match = hmac.compare_digest(self._der_encoded_bytes(), bytes(other))
-        return simple_bytes_match or der_encoded_match
+    # These assumptions simplify the algorithm
+    assert CURVE.cofactor == 1
+    assert CURVE.field_element_size == CURVE.scalar_size
 
+    point_r = CurvePoint.from_bytes(b'\x02' + bytes(r))
+    e = CurveScalar.from_prehashed_for_signing(prehashed_message)
+    inv_r = r.invert()
 
-class Signer:
-    """Callable wrapping ECDSA signing with UmbralPrivateKeys"""
-
-    def __init__(self,
-                 private_key: UmbralPrivateKey,
-                 hash_algorithm: Type[HashAlgorithm] = DEFAULT_HASH_ALGORITHM) -> None:
-        self.__cryptography_private_key = private_key.to_cryptography_privkey()
-        self.curve = private_key.params.curve
-        self.hash_algorithm = hash_algorithm
-
-    def __call__(self, message: bytes, is_prehashed: bool = False) -> Signature:
-        """
-         Signs the message with this instance's private key.
-
-         :param message: Message to hash and sign
-         :param is_prehashed: True if the message has been prehashed previously
-         :return: signature
-         """
-        if is_prehashed:
-            signature_algorithm = ECDSA(utils.Prehashed(self.hash_algorithm()))
-        else:
-            signature_algorithm = ECDSA(self.hash_algorithm())
-
-        signature_der_bytes = self.__cryptography_private_key.sign(message, signature_algorithm)
-        return Signature.from_bytes(signature_der_bytes, der_encoded=True, curve=self.curve)
+    ge = CurvePoint.generator() * e
+    rs = point_r * s
+
+    # Try `r`
+    q = (rs - ge) * inv_r
+    if q == public_key.point():
+        return False
```

Comment:
Noticed this yesterday during the walkthrough - how come this returns False?

Reply:
This function returns a recovery byte, which can only be `\x00` or `\x01`, so I return a boolean flag instead.